### ISSUE TITLE
"A task was canceled." error from RedeemAuthorizationCodeAsync from WebApp

### ISSUE THREAD

**Issue by shrimpy** [OneDrive/onedrive-sdk-dotnet-msa-auth-adapter#6]
With below code, "A task was canceled." was throw when try to authorize token.

````C#
        [HttpGet]
        [Route("api/Auth/OneDriveAuth")]
        public RedirectResult AuthOneDrive()
        {
            OAuthHelper helper = new OAuthHelper();
            return this.Redirect(helper.GetAuthorizationCodeRequestUrl(OneDriveClientId, OneDriveRedirectUrl, new string[] { "offline_access", "onedrive.readonly" }));
        }

        [HttpGet]
        [Route("api/Auth/OneDriveCallback")]
        public async Task<object> AuthOneDriveCallback(string code)
        {
            try
            {
                MyOAuthHelper helper = new MyOAuthHelper();
                AccountSession account = await helper.RedeemAuthorizationCodeAsync(code, OneDriveClientId, OneDriveSecret, OneDriveRedirectUrl, new string[] { "offline_access", "onedrive.readonly" });
                return account;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
````

Updated to below code, i am able to get correct token
````C#
              string body = helper.GetAuthorizationCodeRedemptionRequestBody(code, OneDriveClientId, OneDriveRedirectUrl, new string[] { "offline_access", "onedrive.readonly" }, OneDriveSecret);
                var content = new StringContent(body);
                content.Headers.ContentType = new MediaTypeHeaderValue(Constants.FormUrlEncodedMediaType);
                using (var client = CreateHttpClient())
                using (var response = await client.PostAsync("https://login.live.com/oauth20_token.srf", content))
                {
                    return await response.Content.ReadAsStringAsync();
                }
````

**Comment by chrispatrick** [comment]
I just ran into this same issue. I believe it comes from using the overload of `RedeemAuthorizationCodeAsync` that doesn't take an `HttpProvider`.

It looks like the `HttpProvider` gets disposed before the task has been awaited, which cancels the request.

Easy workaround is to surround the call to `RedeemAuthorizationCodeAsync` with `using(var httpProvider = new HttpProvider())` and pass that `HttpProvider` in.

**Comment by cdmayer** [comment]
This is a bug. Thanks for reporting. Going to send out the fix in the next update, which should be coming soon.

Meanwhile, use @chrispatrick 's workaround.

**Comment by cdmayer** [comment]
@shrimpy I believe this is fixed in 1.0.3. Can you update your NuGet package and give it a shot?

**Comment by cdmayer** [comment]
Fixed
